Porting txstate

devanshj opened the pull request:
Supercedes #36 

I was trying to write some types for implementation for you so I even rewrote the whole thing because I felt like and I was bored ahaha. Take no pressure in merging this but if you like it do it and you can also use the types from here.

There's some really good stuff in here that is branded types and assertNever for exhaustive check and strongly typed record which I don't have time to explain because I'm going to sleep :P but the gist is as all primitive are branded you can't make mistakes like these...
```ts
definition.states[event.type] // nope
definition.states["foo"] // nope
R.get(definition.states, machineState.value) // the only way to access the node
R.get(definition.states, transition.target) // also works till the time the type is the same

let x = machineState.value;
x = event.type // nope both are string in runtime but aren't of the same type so can't be assigned.
```

Haven't checked if the tests are passing or not btw didn't have time was in a rush. Coz really I just rewrote this for fun don't mind me xD (this branch is based on `txstate-port` btw, if you want to quickly check the diff you can do it [here](https://github.com/devanshj/useStateMachine/pull/1)). And if it interests you feel free to ask questions.

And also if you like those `Impl` types let me know if you want me to add them to `txstate-port` branch itself.


To-do:

- [x] Rebase changes from "Porting txstate" branch
- [x] Explicit initial event (https://github.com/cassiozen/useStateMachine/issues/60)
- [x] Fix tests
- [x] Fix issue with guards
- [x] See if we can have types ContextImpl instead of Context.Impl 
- [ ] Update Docs
- [x] Update Examples
- [ ] Devansh's release article
- [ ] Cassiozen's release video
- [x] Enforce usage of `t`
- [ ] Check what happens if the definition is not literal.
- [ ] `lite-typed` version
- [x] `nextEvents.includes` inferred as `never` (https://github.com/cassiozen/useStateMachine/issues/62)

cassiozen commented:
Hey @devanshj - I just wanted to say that this is awesome! I'm on crunch time at work so I didn't have time to review It yet, but I'm super excited to merge!

devanshj commented:
Thanks I'm glad you liked it! I didn't even expect you to want to merge it haha. And no worries at all take your time, I understand.

cassiozen commented:
Ok, I started reading the code. I will probably more questions as I go ;)

1. Is you `extras/R` kind of a tiny TS Rambda of sorts?
1. This version doesn't yet support nested `states`, correct? 
1. Right now, I can't build - it warns about TS errors. It might be a TSDX issue. Can you build in your machine?

Some other observations & things to do:

1. This version doesn't have a specific log for when a transition is blocked by a guard and doesn't log successful transitions. I can work on that.
1. Loved `assertNever`.
1. Need to update examples and failing tests. I can also work on those
1. Need to update docs and work on a release article. Can we split the work on this? Do you have the availability?

Overall I love how this is going. You put a lot of effort into this, so I'd like to invite you to be a maintainer of this project with me. Is this something that interests you?

devanshj commented:
> Is you extras/R kind of a tiny TS Rambda of sorts?

Not really, `R` stands `Record`. It only has functions you'd use on an object used as a map. So instead of `a[k]` you'd write `R.get(a, k)` and instead of `Object.keys(a)` you'd write `R.keys(a)`. The reason is that typescript doesn't allow indexing via branded string, so now `{ [_ in string & { __someTag: true }]: "foo" }` is actually `{ __K: string & { __someTag: true }, __V: "foo" }` instead. Let me know if you want me to elaborate

> This version doesn't yet support nested states, correct?

That's right

> TSDX (the setup tool used in use-statemachine) doesn't support TypeScript 4 yet, but there's a workaround using yarn resolutions - I updated the project to use it, so make sure to use yarn from now on.

Sure

> This version doesn't have a specific log for when a transition is blocked by a guard and doesn't log successful transitions. I can work on that.

Ah must have missed, thanks.

> The minimized bundle is now bigger than half a kb (even using brotli instead of gzip), so maybe we should rename the catch-phrase from "The ½ kb state machine hook for React" to "The <1 kb state machine hook for React" 😂

Haha still has a lot of useless stuff like `assertNever` can be an empty function we need not throw the error it's only because we get compile-time exhaustive checks.

> Loved assertNever.

Haha thanks, idr if it was original or inspired xD

> Need to update examples and failing tests. I can also work on those

Yep that'd be good.

> Need to update docs and work on a release article. Can we split the work on this? Do you have the availability?

Sure thing, I'd prefer to work on a release article. I was thinking of hosting it somewhere (as opposed to md) so we can have twoslash intergration (as opposed to screenshots) but then I think that'd be too much of work :P wdyt?

> You put a lot of effort into this, so I'd like to invite you to be a maintainer of this project with me. Is this something that interests you?

I understand where you're coming from but it's not something that I'd opt into majorly because this is something I do for fun quite literally and maintaining a library is no fun xD Look I understand that contribution comes with an implicit responsibility which I'll gladly fulfil (eg helping you out if users report some ts issues etc) but "maintainer" is a overhead responsibility for me especially given the area this library focuses is not really of my interest, I've never written a state machine nor have I even used xstate, it just happened to be a typing challenge that I took. So thanks for the offer but I don't think I'll be able to take it sorry!

devanshj commented on src/extras.ts:
```diff
@@ -0,0 +1,45 @@
+  return ref.current;
+}
+
+export const assertNever = (value: never): never => {
+  throw new AssertNeverInvariant(value)
+}
+class AssertNeverInvariant extends Error {
+  constructor(public value: never) {
+    super("Invariant: assertNever was called")
+  }
+}
```
```suggestion
export const assertNever = (value: never): never => {}
```

I guess this would make the size go down by quite a margin

devanshj commented:
Other things pending here -
1. rebase to `txstate-port` (there are some changes I did)
2. See if we can have types `ContextImpl` instead of `Context.Impl` because the intellisense quickinfo tooltip shows the type of everything as `Impl` which is hella annoying and useless

devanshj commented on src/index.tsx:
```diff
@@ -1,280 +1,142 @@
-            );
-          return state;
-        }
+      let nextStateValue =
+        typeof resolvedTransition === "string" ? resolvedTransition :
+        resolvedTransition.guard === undefined ? resolvedTransition.target :
+        resolvedTransition.guard({ context, event }) ? resolvedTransition.target :
+        undefined;
+
+      if (nextStateValue === undefined) {
+        log(
+          `Current state doesn"t listen to event type "${event.type}".`,
+          ["Current State", machineState],
+          ["Event", event]
+        )
+        return machineState;
+      }
```
```suggestion
      let selectedTransition =
        typeof resolvedTransition === "string" ? { target: resolvedTransition } :
        resolvedTransition.guard === undefined ? { target: resolvedTransition.target } :
        resolvedTransition.guard({ context, event }) ? { target: resolvedTransition.target, denied: true } :
        undefined;

      if ("denied" in selectedTransition && selectedTransition.denied) {
        log(
          `Transition from "${machineState.value}" to "${selectedTransition}" denied by guard`,
          ["Event", event],
          ["Context", context]
        )
        return machineState;
      }
```

devanshj commented on src/index.tsx:
```diff
@@ -1,280 +1,142 @@
+        )
+        return machineState;
+      }
+      
```
```suggestion
      log(`Transition from "${machineState.value}" to "${nextStateValue}"`, ['Event', eventObject])
      
```

nitpick: maybe rename `nextStateValue` to `nextMachineStateValue` I value consistency and specificity way higher than length of the a variable

cassiozen replied:
Agreed

devanshj commented on src/index.tsx:
```diff
@@ -1,280 +1,142 @@
-}
+const useStateMachine = useStateMachineImpl as UseStateMachine
+export default useStateMachine;
+export const t = <T extends unknown>() => null as T
```
I was thinking of making it 
```ts
export const $$t = Symbol("t");
export const t = <T extends unknown>() => $$t as T;
```
and then add a runtime check if `$$t` is what the users are passing in the schema or not, if not then log a warning that they are supposed to use `t`. But then it's trivial and would add to the size of the library.

cassiozen replied:
Just so we're clear: I don't think library size is the most important metric here. I used it as a catchphrase because it happened to be half a kb, but I don't mind it being bigger.

devanshj replied:
Haha yeah I'm aware. But I'd only suggest to add this if you think I'd be useful, so I'll let you take the call.

cassiozen replied:
I do think it would be useful ;)

devanshj replied:
We should add it then! I think a separate `checkSchema` would be good.

devanshj replied:
Perhaps something like...
```ts
const checkSchema = schema => {
  if (!schema) return;
  checkContextSchema(schema.context);
  checkEventsSchema(schema.events);
}

const checkContextSchema = schema = {
  if (!definition.schema.context) return;
  if (defintion.schema.context === $$t) return;
  warn("whatever")
}

const checkEventsSchema = schema => {
  if (!definition.schema.events) return;
  if (typeof definition.schema.events !== "obejct") {
    warn("...")
    return;
  }
  if (
    R.entries(defintion.schema.events)
    .every(([k, v]) => k === "$$exhuastive" ? true : v === $$t)
  ) return;
  warn("whatever")
}
```

devanshj replied:
Actually you know what I realized we don't need to do this I can enforce people use `t` via types itself. I'll update the types.

devanshj commented on src/index.tsx:
```diff
@@ -1,280 +1,143 @@
+
+      if ("isDenied" in selectedTransition && selectedTransition.isDenied) {
+        log(
+          `Transition from "${machineState.value}" to "${selectedTransition}" denied by guard`,
```
```suggestion
          `Transition from "${machineState.value}" to "${selectedTransition.target}" denied by guard`,
```

Made this edit last time after you committed xD

devanshj commented on src/index.tsx:
```diff
@@ -1,280 +1,143 @@
+
+      if (!resolvedTransition) {
+        log(
+          `Current state doesn"t listen to event type "${event.type}".`,
```
```suggestion
          `Current state doesn't listen to event type "${event.type}".`,
```

devanshj commented on src/extras.ts:
```diff
@@ -0,0 +1,38 @@
+    r2: R2
+  ) =>
+    ({ ...r1, ...r2 }) as any as R.Concat<R1, R2>,
+  fromMaybe: <R extends R.Unknown>(r: R | undefined) => r || {} as R,
```
```suggestion
  fromMaybe: <R extends R.Unknown>(r: R | undefined) => r ?? {} as R,
```

cassiozen commented:
> So thanks for the offer but I don't think I'll be able to take it sorry!

Don't worry, I understand.

> Sure thing, I'd prefer to work on a release article. I was thinking of hosting it somewhere (as opposed to md) so we can have twoslash intergration (as opposed to screenshots) but then I think that'd be too much of work :P wdyt?

I think you should absolutely do it. Is there any public platform that supports twoslash (like dev.to)?

devanshj commented:
> Don't worry, I understand.

Thanks for understanding!

> Is there any public platform that supports twoslash (like dev.to)?

I doubt. We can use gh-pages I guess. I could use [remark-shiki-twoslash](https://www.npmjs.com/package/remark-shiki-twoslash) or something like that to output html.

devanshj commented on src/extras.ts:
```diff
@@ -0,0 +1,38 @@
+  fromMaybe: <R extends R.Unknown>(r: R | undefined) => r || {} as R,
+  keys: <R extends R.Unknown>(r: R) => Object.keys(r) as R.Key<R>[]
+}
+export namespace R {
+  export type Of<K extends string, V> =
+    { __V: V, __K: K }
+  
+  export type Unknown =
+  { __V: unknown, __K: string }
+  
+  export type Key<R extends R.Unknown> =
+    R extends { __K: infer K } ? K : never;
+  export type Value<R extends R.Unknown> =
+    R extends { __V: infer V } ? V : never
+  export type Concat<R1 extends R.Unknown, R2 extends R.Unknown> =
+    R.Of<
+      R.Key<R1> | R.Key<R2>,
+      R.Value<R1> | R.Value<R2>
+    >  
+}
```
```suggestion
export namespace R {
  export type Of<K extends string, V> = { [$$K]: K, [$$V]: V }
  export declare const $$K: unique symbol;
  export declare const $$V: unique symbol;
  export type Unknown = Of<string, unknown>

  export type Key<R extends R.Unknown> = R[$$K]
  export type Value<R extends R.Unknown> = R[$$V]
  export type Concat<R1 extends R.Unknown, R2 extends R.Unknown> =
    R.Of<
      R.Key<R1> | R.Key<R2>,
      R.Value<R1> | R.Value<R2>
    >  
}
```

For the sake of restricting `states.__K` and `states.__V` access.

cassiozen replied:
I reverted this because it caused an error: Identifier 'R' has already been declared

devanshj replied:
Fixed it!

cassiozen commented:
Question: Prettier is screaming at me when I open the files. I see that you like to be precise about the formatting - can you live with some configuration of Prettier? We can disable it too, but I would prefer to keep it.

devanshj commented:
> I see that you like to be precise about the formatting

Good observation :P I was about to mention about prettier, I tell you what -- I absolutely hate it. I could write an essay on how it's just wrong. But hey it's your project if you prefer to keep it do that. Though I doubt it'd matter as it's not like a big org where people would have diff opinion about formatting and would bike-shed all the time (that's the only place I can give prettier a benefit of doubt). Though again I'd have you take the call.

devanshj commented:
Did you rebase to the latest `txstate-port`?

cassiozen commented:
>  I tell you what -- I absolutely hate it. I could write an essay on how it's just wrong

🤣😆

> I'd have you take the call. (...) I would still want prettier to ignore `types.ts`

Deal - I'll keep ignoring `types.ts` and I left the prettier requirements less strict, but I'll keep on index and extra.

devanshj commented on src/index.tsx:
```diff
@@ -81,63 +67,59 @@ const createReducer = (definition: Machine.Definition.Impl) => {
-        resolvedTransition.guard === undefined ? { target: resolvedTransition.target } :
-        resolvedTransition.guard({ context, event }) ? { target: resolvedTransition.target } :
-        { target: resolvedTransition.target, isDenied: true };
+        typeof resolvedTransition === "string"
+          ? { target: resolvedTransition }
+          : resolvedTransition.guard === undefined
+          ? { target: resolvedTransition.target }
+          : resolvedTransition.guard({ context, event })
+          ? { target: resolvedTransition.target }
+          : { target: resolvedTransition.target, isDenied: true };
```
This is what I was talking about absolute dogshit 😂 Feel free to refactor into a non-ternary version because now it's no longer readable to me at least

cassiozen commented:
I'm having a problem with guards on this branch: Whenever I add a guard to a transition, its expected type is undefined:

<img width="878" alt="guard_undefined" src="https://user-images.githubusercontent.com/33676/126901152-8d24e392-f90e-4751-a798-25bbd3eb1d8e.png">

cassiozen commented:
> Did you rebase to the latest `txstate-port`?

I did not

cassiozen commented:
I edited the first comment to add a bunch of to-do tasks - hope you don't mind.

devanshj commented:
> I'm having a problem with guards on this branch: Whenever I add a guard to a transition, its expected type is undefined:

Yeah I noticed and fixed it on `txstate-port` (I'll rebase to that). Either I missed it or maybe regression from 4.3.2 to 4.3.5 not sure

devanshj commented:
> I edited the first comment to add a bunch of to-do tasks - hope you don't mind.

No that's great we'd have clarity!

devanshj commented:
I wanted to keep `txstate-port` to as source of truth for `types.ts` so I cherry-picked the type changes from `rewrite` (the `Impl` types etc) to `txstate-port` then rebased `rewrite` to `txstate-port` and finally force pushed, hopefully I didn't mess up anything :P

devanshj commented on src/index.ts:
```diff
@@ -0,0 +1,134 @@
+  if (!definition.verbose) return;
+  
+  let console = definition.console || defaultConsole
+  if (process.env.NODE_ENV === "development" || process.env.NODE_ENV === "test") {
```
Can we remove this? Not sure what's the point, we already have `verbose`. If someone passed `verbose: true` for prod, it's their fault :P

```suggestion
```

cassiozen replied:
Oh, this signals babel that it's safe to tree shake this bit of code. We get a few extra bites smalles.

devanshj replied:
Ah nice in that case I should wrap the whole body of the function in `if`?

cassiozen commented:
Hey @devanshj - Do you think this is in a state we could release another beta before the official 1.0 release? It could be nice to have some additional user testing.

cassiozen commented:
Found a bug while converting the timer example. Filled separately for better organization: https://github.com/cassiozen/useStateMachine/issues/62

devanshj commented:
> Do you think this is in a state we could release another beta before the official 1.0 release? It could be nice to have some additional user testing.

Sure thing but not yet, I'll mark this a draft for clarity and will open once it's ready or at least ready enough for a beta. Right now enforcing `t` and disallowing `$$initial` as an event type are remaining then we'd be ready for another beta.

devanshj commented on tsconfig.json:
```diff
@@ -1,33 +1,22 @@
-    // match output dir to input dir. e.g. dist/index instead of dist/src/index
-    "rootDir": "./src",
-    // stricter type-checking for stronger correctness. Recommended by TS
+    // "rootDir": "./src",
```
(from txstate-port) I think removing `"rootDir": "./src"` will affect the build (?), but I want to type-check "types" folder too. How to go about this?

devanshj commented:
@cassiozen I think this is good for another beta.

cassiozen commented:
> I think this is good for another beta

Awesome. I created a new branch `1.0.0-beta` and rebased this PR to merge and bake a new release from the original repo.

We can continue working on this same PR on your fork, or open a new PR, whatever you prefer.

cassiozen commented:
@all-contributors please add @devanshj  for bug, code, ideas and test

devanshj commented:
The only thing pending other than documentation is lite-types, I think I'll open a new PR for that and base it on the beta branch. You can add commits changing documentation on the beta branch directly if you want.
